shared-bindings/ps2io/Ps2.c

```diff
@@ -0,0 +1,194 @@
+        { MP_QSTR_clkpin, MP_ARG_REQUIRED | MP_ARG_OBJ },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    assert_pin(args[ARG_clkpin].u_obj, false);
+    assert_pin(args[ARG_datapin].u_obj, false);
+    const mcu_pin_obj_t* clkpin = MP_OBJ_TO_PTR(args[ARG_clkpin].u_obj);
+    assert_pin_free(clkpin);
+    const mcu_pin_obj_t* datapin = MP_OBJ_TO_PTR(args[ARG_datapin].u_obj);
+    assert_pin_free(datapin);
+
+    ps2io_ps2_obj_t *self = m_new_obj(ps2io_ps2_obj_t);
+    self->base.type = &ps2io_ps2_type;
+
+    common_hal_ps2io_ps2_construct(self, datapin, clkpin);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|   .. method:: deinit()
+//|
+//|      Deinitialises the Ps2 and releases any hardware resources for reuse.
+//|
+STATIC mp_obj_t ps2io_ps2_deinit(mp_obj_t self_in) {
+    ps2io_ps2_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_ps2io_ps2_deinit(self);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(ps2io_ps2_deinit_obj, ps2io_ps2_deinit);
+
+//|   .. method:: __enter__()
+//|
+//|      No-op used by Context Managers.
+//|
+//  Provided by context manager helper.
+
+//|   .. method:: __exit__()
+//|
+//|      Automatically deinitializes the hardware when exiting a context. See
+//|      :ref:`lifetime-and-contextmanagers` for more info.
+//|
+STATIC mp_obj_t ps2io_ps2_obj___exit__(size_t n_args, const mp_obj_t *args) {
+    (void)n_args;
+    common_hal_ps2io_ps2_deinit(args[0]);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(ps2io_ps2___exit___obj, 4, 4, ps2io_ps2_obj___exit__);
+
+//|   .. method:: get_byte()
+//|
+//|     Removes and returns the oldest received byte. When buffer
+//|     is empty, returns a negative number.
+//|
+STATIC mp_obj_t ps2io_ps2_obj_get_byte(mp_obj_t self_in) {
+    ps2io_ps2_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    raise_error_if_deinited(common_hal_ps2io_ps2_deinited(self));
+
+    return MP_OBJ_NEW_SMALL_INT(common_hal_ps2io_ps2_get_byte(self));
+}
+MP_DEFINE_CONST_FUN_OBJ_1(ps2io_ps2_get_byte_obj, ps2io_ps2_obj_get_byte);
+
+//|   .. method:: send_byte()
+//|
+//|     Sends a byte to PS/2. Returns the response byte (positive)
+//|     or a communication error code (negative). The response is
+//|     typically an ACK (0xFA). Some commands return additional data
+//|     bytes that should be received via get_byte().
+//|
+STATIC mp_obj_t ps2io_ps2_obj_send_byte(mp_obj_t self_in, mp_obj_t ob) {
+    ps2io_ps2_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    raise_error_if_deinited(common_hal_ps2io_ps2_deinited(self));
+    mp_int_t b = mp_obj_get_int(ob) & 0xff;
+    return MP_OBJ_NEW_SMALL_INT(common_hal_ps2io_ps2_send_byte(self, b));
+}
+MP_DEFINE_CONST_FUN_OBJ_2(ps2io_ps2_send_byte_obj, ps2io_ps2_obj_send_byte);
+
+//|   .. method:: get_error()
+//|
+//|     Returns and clears the bitmap of last communication errors.
```

Comment:
Where is this bitmap defined? Is it a PS/2 thing? Is it per device?

Reply:
No, it was arbitrary. More a (hardware) debug tool than anything else.

Reply:
Ok, the pythonic way would be to have different exceptions for each case. Its a bit tricky to define new exceptions in C by the gist is here: https://github.com/adafruit/circuitpython/blob/master/py/objexcept.c#L297 It would be done at the ps2io module level so that they are only built in when the module is.

Reply:
The rx errors at least arise asynchronously, so no opportunity to raise exceptions, and the error cannot be "handled", so I think moving TX errors to clear_errors() and just report TX errors as RuntimeError's is best.